### DISCUSSION TITLE
What happens to atomWithStorage items when removed? Can you migrate?

### DISCUSSION THREAD

**Discussion by GollyJer** [pmndrs/jotai#921] [Q&A]
I'm eager to try out jotai but curious about persisted data changes.

What happens when I remove an `atomWithStorage` atom from the code, or rename one? Is the persisted data stuck on disk?

Am I missing some way to migrate persisted data?

Am I stuck in a flux mindset and need to change the way I'm thinking this?

Thanks!

**Comment by dai-shi** [comment]
There's no magic behind `atomWithStorage`. If you remove it from the code, the data will be kept in localStorage.
For migration, you would need to provide a custom `storage`. We don't yet have a detailed doc about it. Basically, you need to implement a SyncStorage, if you are using localStorage.

```ts
type SyncStorage<Value> = {
  getItem: (key: string) => Value
  setItem: (key: string, newValue: Value) => void
}
```

**Reply by GollyJer** [reply]
Thanks for the response @dai-shi. We are currently using React Native with easy-peasy for state management. It's been great but there isn't any upkeep anymore.

I really want to give atom based state a try but for now I think we'll go with Zustand until there's a clear path for persistence and migration which you already have over there.

If I did want to tackle a custom storage (which I actually have built already for an Expo based project) how would I plug it into jotai?

**Reply by GollyJer** [reply]
Or maybe just set up [Zustand integration](https://github.com/pmndrs/jotai/blob/main/docs/integrations/zustand.mdx) to hold the store and use it's persistence functionality?

**Reply by dai-shi** [reply]
The custom `storage` I mean is the third argument of `atomWithStorage`.

https://github.com/pmndrs/jotai/blob/750db553dedadc505995d75b19b884c6bc4112a3/src/utils/atomWithStorage.ts#L65

It will be `type AsyncStorage` (not to confuse with RN's AsyncStorage).

If you don't need migration, the easiest way to create it is `const storage = createJSONStorage(() => AsyncStorage)`. (this is RN's AsyncStorage.)

(and, we are looking for someone to help improve docs around this. #805)

**Reply by dai-shi** [reply]
I might have misunderstood what you mean by migration.

both,
easy-peasy -> zustand
easy-peasy - > jotai
should be fine.

No need to go,
easy-peasy -> zustand -> jotai

**Reply by GollyJer** [reply]
Your original interpretation is correct.  This will be a rewrite from easy-peasy to jotai. 

I'm focused on the functionality I don't want to lose which is the ability to migrate the local data when we change it's structure.

Here's an exact example we often do...
Our users have a list of addresses (in the store) and all the other data is queried and organized based of these addresses.
Often we update how the other data is structured but don't want to lose the addresses.
To do this we "wipe" all the state except the addresses.
```
const migrations = {
  0: (state) => ({ addresses: { ...state.addresses } })
  1: (state) => ({ addresses: { ...state.addresses } }),
}

const persistConfig = {
  storage: ExpoFileSystemStorage,
  version: 1, // Change this to match the migration number.
  migrate: createMigrate(migrations),
};
```
After a migration the app pulls down all the data for each address and puts the results in the new structure.
Without this ability our users would need to re-enter their addresses; we don't want to make them do this.

I think you've answered how I can change the storage to persist for React Native but I'm missing how a migration like this might work.

Thank you for taking the time to help!!

**Reply by GollyJer** [reply]
Should I be looking at valtio instead of jotai? 
Seems like it might be even better.

Sorry, I'm trying to get up to speed on all these amazing new ways of managing state.

**Reply by dai-shi** [reply]
The normal `atomWithStorage` usage in RN would be:
```js
import { atomWithStorage, createJSONStorage } from 'jotai/utils'

const storage = createJSONStorage(() => AsyncStorage)
const fooAtom = atomWithStorage('foo', null, storage)
```

Let's suppose you have a `migrate` function that you can call even if it's already the latest format.
You could do something like this:
```ts
import { atomWithStorage, createJSONStorage } from 'jotai/utils'

const baseStorage = createJSONStorage(() => AsyncStorage)
const storage: typeof baseStorage = {
  ...baseStorage,
  getItem: async (key) => {
    const value = await baseStorage.getItem(key)
    return migrate(value)
  },
}
const fooAtom = atomWithStorage('foo', null, storage)
```

**Reply by dai-shi** [reply]
> Should I be looking at valtio instead of jotai?

valtio is close to zustand in that it's module states.
jotai is modeled after useContext+useState.
If you need to read store values outside React, valtio and zustand would be easier.
